File: neon_tetris_backend/src/models/user_model.js
import mongoose from "mongoose";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";
import mongooseAggregatePaginate from "mongoose-aggregate-paginate-v2";

const userSchema = new mongoose.Schema(
  {
    username: {
      type: String,
      required: true,
      trim: true,
    },
    email: {
      type: String,
      required: true,
      unique: true,
      lowercase: true,
      index: true,
    },
    password: {
      type: String,
      required: [true, "Password cannot be empty"],
    },
    avatarUrl: {
      type: String,
      default:
        "https://static.vecteezy.com/system/resources/previews/007/407/996/non_2x/user-icon-person-icon-client-symbol-login-head-sign-icon-design-vector.jpg",
    },
    friends: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],
    rankedStats: {
      gamesPlayed: { type: Number, default: 0 },
      gamesWon: { type: Number, default: 0 },
      linesCleared: { type: Number, default: 0 },
      highestScore: { type: Number, default: 0 },
    },
    unrankedStats: {
      gamesPlayed: { type: Number, default: 0 },
      gamesWon: { type: Number, default: 0 },
      linesCleared: { type: Number, default: 0 },
      highestScore: { type: Number, default: 0 },
    },
    isOnline: {
      type: Boolean,
      default: false,
    },
    lastLogin: {
      type: Date,
    },
    refreshToken: {
      type: String,
    },
  },
  { timestamps: true }
);

// Hash password before save
userSchema.pre("save", async function (next) {
  if (!this.isModified("password")) return next();
  this.password = await bcrypt.hash(this.password, 10);
  next();
});

// Compare passwords
userSchema.methods.isPasswordCorrect = async function (password) {
  return await bcrypt.compare(password, this.password);
};

// JWT methods
userSchema.methods.generateAccessToken = function () {
  return jwt.sign(
    { _id: this._id, email: this.email },
    process.env.ACCESS_TOKEN_SECRET,
    { expiresIn: process.env.ACCESS_TOKEN_EXPIRY }
  );
};

userSchema.methods.generateRefreshToken = function () {
  return jwt.sign({ _id: this._id }, process.env.REFRESH_TOKEN_SECRET, {
    expiresIn: process.env.REFRESH_TOKEN_EXPIRY,
  });
};

userSchema.plugin(mongooseAggregatePaginate);
export const User = mongoose.model("User", userSchema);

File: neon_tetris_frontend/lib/models/user_model.dart
import 'package:equatable/equatable.dart';

class User extends Equatable {
  final String id;
  final String username;
  final String? email;
  final String? avatarUrl;
  final List<String>? friends;
  final GameStats? rankedStats;
  final GameStats? unrankedStats;
  final bool? isOnline;
  final DateTime? lastLogin;

  const User({
    required this.id,
    required this.username,
    this.email,
    this.avatarUrl,
    this.friends,
    this.rankedStats,
    this.unrankedStats,
    this.isOnline,
    this.lastLogin,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    final userData = json['user'] ?? json; // Handles nested or flat user data
    return User(
      id: userData['_id'] ?? userData['userId'],
      username: userData['username'],
      email: userData['email'],
      avatarUrl: userData['avatarUrl'],
      friends: List<String>.from(userData['friends'] ?? []),
      rankedStats: GameStats.fromJson(userData['rankedStats'] ?? {}),
      unrankedStats: GameStats.fromJson(userData['unrankedStats'] ?? {}),
      isOnline: userData['isOnline'] ?? false,
      lastLogin: userData['lastLogin'] != null
          ? DateTime.parse(userData['lastLogin'])
          : null,
    );
  }

  @override
  List<Object?> get props => [
        id,
        username,
        email,
        avatarUrl,
        friends,
        rankedStats,
        unrankedStats,
        isOnline,
        lastLogin,
      ];
}

class GameStats extends Equatable {
  final int gamesPlayed;
  final int gamesWon;
  final int linesCleared;
  final int highestScore;

  const GameStats({
    this.gamesPlayed = 0,
    this.gamesWon = 0,
    this.linesCleared = 0,
    this.highestScore = 0,
  });

  factory GameStats.fromJson(Map<String, dynamic> json) {
    return GameStats(
      gamesPlayed: json['gamesPlayed'] ?? 0,
      gamesWon: json['gamesWon'] ?? 0,
      linesCleared: json['linesCleared'] ?? 0,
      highestScore: json['highestScore'] ?? 0,
    );
  }

  @override
  List<Object> get props => [gamesPlayed, gamesWon, linesCleared, highestScore];
}